### DISCUSSION TITLE
Variants

### DISCUSSION THREAD

**Discussion by jsirish** [dynamic/silverstripe-shopify#18] [Ideas]
ClassName: ShopifyVariant extends DataObject

Products may have multiple variations, such as size, color, etc. Variants are managed in the product records in the Shopify Store Admin. Variants can have a custom price, sku, image, and amount available.

Because we are generating the add to cart form via the Buy Button, there's a chance we wouldn't need to sync this data. However, all pricing information for Products live on Variants, so we will import them to calculate a display price.

ShopifyVariant fields to be mapped from Shopify:

- $db
  - title => Title (Varchar)
  - sku => SKU (Varchar)
  - price => Price (Currency)
  - compare_at_price => ComparePrice (Currency)
  - created_at => Created (DateTime)
  - updated_at => LastEdited (DateTime)
  - id => ShopifyID (Int)
- $has_one
  - image_id => File (ShopifyFile)
  - product => Product (ShopifyProduct)

[Full field list](https://shopify.dev/docs/admin-api/rest/reference/products/product-variant).

**Comment by jsirish** [comment]
@muskie9 @mak001

**Comment by mak001** [comment]
If we don't have sku on the base product we should generate a default variant when a product is created. I think it should also include the position field.

**Comment by jsirish** [comment]
I think that's a good idea. We considered doing something similar for Foxy. Since all Product-related records will be read-only CMS UX isn't a major concern here.

**Reply by jsirish** [reply]
The current import task does create a default variant. I'm able to query the first Variant to get such data as the SKU
